share/vss/vss.go

```diff
@@ -0,0 +1,630 @@
+	// library.
+	Signature []byte
+}
+
+const (
+	// StatusComplaint is a constant value meaning that a verifier issues
+	// a Complaint against its Dealer.
+	StatusComplaint byte = iota
+	// StatusApproval is a constant value meaning that a verifier agrees with
+	// the share it received.
+	StatusApproval
+	// special status when a complaint has been justified
+	statusJustified
+)
+
+// Justification is a message that is broadcasted by the Dealer in response to
+// a Complaint. It contains the original Complaint as well as the shares
+// distributed to the complainer.
+type Justification struct {
+	// SessionID related to the current run of the protocol
+	SessionID []byte
+	// Index of the verifier who issued the Complaint,i.e. index of this Deal
+	Index uint32
+	// Deal in clear text
+	Deal *Deal
+	// Signature over the whole packet. It is automatically verified by this
+	// library.
+	Signature []byte
+}
+
+// NewDealer returns a Dealer capable of leading the secret sharing scheme. It
+// does not have to be trusted by other Verifiers. The security parameter t is
+// the number of shares required to reconstruct the secret. It is HIGHLY
+// RECOMMENDED to use a threshold higher or equal than what the method
+// MinimumT() returns, otherwise it breaks the security assumptions of the whole
+// scheme. It returns an error if the t is inferior or equal to 2 or if it is
+// not inferior to t.
+func NewDealer(suite abstract.Suite, longterm, secret abstract.Scalar, verifiers []abstract.Point, r cipher.Stream, t int) (*Dealer, error) {
+	d := &Dealer{
+		suite:     suite,
+		long:      longterm,
+		secret:    secret,
+		verifiers: verifiers,
+	}
+	if !validT(t, verifiers) {
+		return nil, fmt.Errorf("dealer: t %d invalid", t)
+	}
+
+	d.t = t
+
+	H := deriveH(d.suite, d.verifiers)
+	f := share.NewPriPoly(d.suite, d.t, d.secret, r)
+	g := share.NewPriPoly(d.suite, d.t, nil, r)
+	d.pub = d.suite.Point().Mul(nil, d.long)
+
+	// F = coeff * B
+	F := f.Commit(d.suite.Point().Base())
+	_, d.secretCommits = F.Info()
+	// G = coeff * H
+	G := g.Commit(H)
+
+	C, err := F.Add(G)
+	if err != nil {
+		return nil, err
+	}
+	_, commitments := C.Info()
+
+	d.sessionID, err = sessionID(d.suite, d.pub, d.verifiers, commitments, d.t)
+	if err != nil {
+		return nil, err
+	}
+
+	d.aggregator = newAggregator(d.suite, d.pub, d.verifiers, commitments, d.t, d.sessionID)
+	// C = F + G
+	d.deals = make([]*Deal, len(d.verifiers))
+	for i := range d.verifiers {
+		fi := f.Eval(i)
+		gi := g.Eval(i)
+		d.deals[i] = &Deal{
+			SessionID:   d.sessionID,
+			SecShare:    fi,
+			RndShare:    gi,
+			Commitments: commitments,
+			T:           uint32(d.t),
+		}
+		if d.deals[i].Signature, err = sign.Schnorr(d.suite, d.long, msgDeal(d.deals[i])); err != nil {
+			return nil, err
+		}
+	}
+	return d, nil
+}
+
+// Deals returns the list of previously generated deals.
+func (d *Dealer) Deals() []*Deal {
+	out := make([]*Deal, len(d.deals))
+	copy(out, d.deals)
+	return out
```

Comment:
Hmm - this doesn't do a deep copy, so if the user modifies the returned value, he will modify the underlying structure. So why do the shallow copy here?